SimpleSAMLphp SP and IdP proxy

Sascha-Peter opened the pull request:
* Add saml salt configuration
* Secure sensitive information with pillar variables

LukeCarrier commented:
@Sascha-Peter I'm _finally_ planning on giving PR a thorough review and test toward the end of next week.

Are you happy for me to rebase this code atop the latest changes? I've changed quite a few of the state names and reworked the service restart behaviour to use `cmd.run` and `onchange` rather than relying on `mod_watch` -- should help us fix the reload vs restart quirks when issuing SSL certs and such.

Sascha-Peter commented:
Sure, please go ahead! That sounds very good about restarts / reloads

On 19 Jul 2017 00:00, "Luke Carrier" <notifications@github.com> wrote:

> @Sascha-Peter <https://github.com/sascha-peter> I'm *finally* planning on
> giving PR a thorough review and test toward the end of next week.
>
> Are you happy for me to rebase this code atop the latest changes? I've
> changed quite a few of the state names and reworked the service restart
> behaviour to use cmd.run and onchange rather than relying on mod_watch --
> should help us fix the reload vs restart quirks when issuing SSL certs and
> such.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LukeCarrier/moodle-ubiquitous/pull/14#issuecomment-316210869>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACxvRnuMACn1AG3JXB0UNr_gizjgLMmRks5sPSrmgaJpZM4Nn8aH>
> .
>

LukeCarrier commented on redis-base/init.sls:
```diff
@@ -0,0 +1,49 @@
+#
```
Please relocate this file to `/redis/init.sls` unless we're adding separate master and slave roles. Since the entire configuration file is being templated and we've opted to go with Sentinel for the clustering configuration I don't think we'll need to do that :)

Sascha-Peter replied:
Done and renamed all imports and documentation references.

LukeCarrier commented on redis-base/redis.conf.jinja:
```diff
@@ -0,0 +1,948 @@
+# Redis configuration file example.
```
Please relocate this file to `/redis/init.sls` unless we're adding separate master and slave roles.

Sascha-Peter replied:
Makes sense. Done.

LukeCarrier commented on redis-base/redis.conf.jinja:
```diff
@@ -0,0 +1,948 @@
+# Redis configuration file example.
```
Can we reduce the comments down in this file to just those that're relevant to our configuration?

Sascha-Peter replied:
I have removed all comments which we haven't had an option for. Left in the ones which we configured.

LukeCarrier commented on _vagrant/salt/pillar/top.sls:
```diff
@@ -4,9 +4,9 @@ base:
     - platforms-logos
 
-  'roles:app':
+  'roles:app-base':
```
I would rather we did not directly assign `*-base` roles to minions. Could this be changed to `roles:app-.+` with a `grains_pcre` match, similar to the Selenium configuration below?

Sascha-Peter replied:
I think this has been resolved?

LukeCarrier commented on _vagrant/salt/pillar/top.sls:
```diff
@@ -4,9 +4,9 @@ base:
+  'roles:app-base':
     - match: grain
-    - app
+    - app-base
```
This change looks incorrect to me, as the corresponding pillar file (`/_vagrant/salt/pillar/app.sls`) was not renamed. Could we reverse this change?

Sascha-Peter replied:
I think this was already resolved too.

LukeCarrier commented on app-base/init.sls:
```diff
@@ -246,187 +253,87 @@ app.{{ domain }}.home:
     - user: www-data
     - group: adm
-    - mode: 0750
+    - mode: 0640
```
This looks like an incorrect merge -- I corrected this in 938f66ab4eaa6d813e9b6c76c18da122e76338a5. Can we incorporate the fix please?

Sascha-Peter replied:
Done. Sorry for that.

LukeCarrier commented on app-base/init.sls:
```diff
@@ -246,187 +253,87 @@ app.{{ domain }}.home:
-{% if pillar['systemd']['apply'] %}
-    - onchanges_in:
-      - app.nginx.reload
-{% endif %}
```
I understand why we're templating the nginx `server {}` blocks (virtual hosts) separately, but is there a reason we're not applying the extra configuration files here?

Sascha-Peter replied:
I think it got confused with the rename. Will add it back in. Don't see a reason not to.

LukeCarrier commented on app-base/init.sls:
```diff
@@ -246,187 +253,87 @@ app.{{ domain }}.home:
     - context:
       domain: {{ domain }}
       instance: blue
+      platform: {{ platform }}
```
This looks like something we missed during our rebasing (see bb829b9d1010d56f11890487f7550ab8b2b76dae -- we moved the `platform` assignment to inside the template).

Sascha-Peter replied:
Yep, sorry, didn't seen that one. Removed.

LukeCarrier commented on redis-base/redis.conf.jinja:
```diff
@@ -0,0 +1,948 @@
+
+# By default Redis does not run as a daemon. Use 'yes' if you need it.
+# Note that Redis will write a pid file in /var/run/redis.pid when daemonized.
+daemonize {{ pillar['redis']['config']['daemonize'] }}
```
Expressing these boolean values as strings in the pillar makes the configuration a little cumbersome. Could we handle this as we do with OpenSSH (see b0c5ca33e50905249ff2e0831f5ce23148a48b59)?

Sascha-Peter replied:
I have adopted that and put it in.

LukeCarrier commented on app-base/init.sls:
```diff
@@ -246,187 +253,87 @@ app.{{ domain }}.home:
-    - onchanges_in:
-      - app.php-fpm.reload
+    - require_in:
+      - service: php-fpm.reload
```
Looks like another rebase issue -- `onchanges_in` is the correct variant of the service restart logic.

LukeCarrier commented on app-base/init.sls:
```diff
@@ -246,187 +253,87 @@ app.{{ domain }}.home:
-    - name: systemctl restart php7.0-fpm
-    - require:
-      - php.packages
-{% endif %}
```
`app.php-fpm.restart` doesn't appear in this file but is referenced as a requisite above. It'll need restoring here, along with `app.nginx.restart` when the `sites-extra` configuration is restored.

Sascha-Peter replied:
Added back in.

LukeCarrier commented on app-base/php-fpm/platform.conf.jinja:
```diff
@@ -27,7 +27,7 @@ ping.path = /ping
 
 catch_workers_output = no
-php_flag[display_errors] = Off
+php_flag[display_errors] = {{ salt['pillar.get']('platform:php:fpm:display_errors', 'Off') }}
```
Couple of things here:
* This is a boolean option, so could we treat it as a boolean in the pillar and turn it into the `"On"`/`"Off"` value with a Jinja macro?
* Do we need this at all? The `app-debug` role already alters this option.

Sascha-Peter replied:
Removed.

LukeCarrier commented on _docs/roles/salt.md:
````diff
@@ -10,7 +10,7 @@ Salt is used as a configuration management system, converging the state of all r
 ```
 roles:
-  - app
+  - app-base
````
`*-base` states should never be directly applied to servers. Please replace with either `app-saml` or `app-moodle`.

Sascha-Peter replied:
Adjusted. Thanks for pointing that one out.

LukeCarrier commented on app-lets-encrypt/init.sls:
```diff
@@ -0,0 +1,53 @@
+app-lets-encrypt.nginx-post-certbot:
+  service.running:
+    - name: nginx
+    - reload: true
```
We need to rework these service restarts as per the `app` role on master, as they likely won't restart the service correctly:
* Salt "magically" replaces `service.running` states with `reload: True` _only if they have a `watch` requisite_.
* A reload is not sufficient to enable SSL in the case where nginx is not already listening on 443 as it doesn't allow nginx to bind to new ports -- it has to be done with a restart.

Sascha-Peter replied:
Reworked those to use the `cmd.run` and use `systemctl`. Please have a look if I understood the issue correctly.

LukeCarrier commented on app-lets-encrypt/init.sls:
```diff
@@ -0,0 +1,53 @@
+#
+
+include:
+  - app-base
```
Since the nginx virtual hosts aren't created in `app-base`, this include won't ensure that they exist. I'm not sure how we can handle this case yet.

Sascha-Peter replied:
Hm... would it be an idea to put it into `nginx-base` as that's included by `app-base` or do we shoot ourselves into the foot with that?

LukeCarrier replied:
It still wouldn't help I'm afraid -- the issue then would be that neither `app-base` nor `nginx-base` would create the necessary configuration files. I think the tidiest solution to this would be moving the per-site configuration into a macro inside `app-lets-encrypt` that we call somewhere inside the `app-moodle` and `app-saml` roles.

Happy to leave this point for now and sit and work on this together at some point this week.

LukeCarrier commented on app-moodle/init.sls:
```diff
@@ -0,0 +1,71 @@
+{% if pillar['systemd']['apply'] %}
+    - require:
+      - pkg: nginx
+{% endif %}
```
These requisites are definitely wrong -- see [master](https://github.com/LukeCarrier/moodle-ubiquitous/blob/master/app/init.sls#L323-L339):
* There's a missing nginx restart here, which should be within the condition
* The dependency on the nginx package has been moved within the condition

Sascha-Peter replied:
My branch seem to be ok, please let me know in case this is not fixed after this review.

LukeCarrier commented on app-moodle/init.sls:
```diff
@@ -0,0 +1,71 @@
+{% if pillar['systemd']['apply'] %}
+    - require_in:
+      - service: nginx.reload
+{% endif %}
```
This should be using `onchanges_in`, see [master](https://github.com/LukeCarrier/moodle-ubiquitous/blob/master/app/init.sls#L341-L350).

Sascha-Peter replied:
Changed already on my branch - let me know if this issue persists.

LukeCarrier commented on app-moodle/init.sls:
```diff
@@ -0,0 +1,71 @@
+    - user: {{ platform['user']['name'] }}
+    - group: {{ platform['user']['name'] }}
+    - mode: 0660
+{% endfor %}
```
The new service restart logic (e.g. [master](https://github.com/LukeCarrier/moodle-ubiquitous/blob/master/app/init.sls#L405-L432)) will need to be added here, as there are missing restarts for both nginx and PHP-FPM. We should need just the restarts here -- enabling the services should take place in `app-base`.

Sascha-Peter replied:
Added both restarts.

LukeCarrier commented on app-moodle/init.sls:
```diff
@@ -0,0 +1,71 @@
+    - user: {{ platform['user']['name'] }}
+    - group: {{ platform['user']['name'] }}
+    - mode: 0660
+{% endfor %}
```
As far as I can see the extra nginx configuration is not applied here either, which will break our ability to manage maintenance pages and such. Please ensure this is added to `app-base` as per the comment there.

Sascha-Peter replied:
I've added that one as per comment further up. Will go through again but hope it's alright.

LukeCarrier commented on app-moodle/init.sls:
```diff
@@ -0,0 +1,71 @@
+    - group: {{ platform['user']['name'] }}
+    - mode: 0770
+    - require:
+      - file: moodle.{{ domain }}.home
```
The [`localcache` state](https://github.com/LukeCarrier/moodle-ubiquitous/blob/master/app/init.sls#L307-L314) will need to be added here -- looks like it got lost during the rebase.

Sascha-Peter replied:
Added. Thanks for pointing this out - sorry for missing it.

LukeCarrier replied:
:+1: that's what reviews are for!

LukeCarrier commented on app-saml/init.sls:
```diff
@@ -0,0 +1,219 @@
+  pkg.installed:
+  - pkgs:
+    - libxml2
+    - libxml2-dev
```
Why do we need these header packages? This package is a dependency of `php-xml` and shouldn't need to be manually installed.

Sascha-Peter replied:
Removed.

LukeCarrier commented on app-saml/init.sls:
```diff
@@ -0,0 +1,219 @@
+    - libxml2
+    - libxml2-dev
+    - zlib1g
+    - zlib1g-dev
```
Why do we need these header packages? This package is a dependency of `php-zip` and shouldn't need to be manually installed.

Sascha-Peter replied:
Removed.

LukeCarrier commented on app-saml/init.sls:
```diff
@@ -0,0 +1,219 @@
+    - openssl
+    - php7.0-sqlite3
+    - php7.0-gmp
+    - libgmp-dev
```
Why do we need these header packages? This package is a dependency of `php-gmp` and shouldn't need to be manually installed.

Sascha-Peter replied:
Removed.

LukeCarrier commented on app-saml/init.sls:
```diff
@@ -0,0 +1,219 @@
+#   file.rename:
+#     - name: {{ platform['user']['home'] }}/simplesamlphp
+#     - source: {{ platform['user']['home'] }}/simplesamlphp-1.14.14
+#     - force: true
```
Please remove the commented out states. As discussed, we'll handle the installation of the release in one of two ways depending on the environment:
* In local development we can use `app-default-release`, which configures the `current` link and sets up the configuration links. With Vagrant we can then rsync a SimpleSAMLphp installation to `~/releases/vagrant` for it to configure for us.
* In production we'll use GoCD, which will need amendments to the GoCD configuration to suit multiple platforms.

Sascha-Peter replied:
Removed.

LukeCarrier commented on _docs/README.md:
```diff
@@ -2,7 +2,7 @@
 Ubiquitous is a multi-purpose Moodle hosting environment.
 
-It fulfils three requirements:
+It fulfils four requirements:
```
Still three requirements, one for each environment ;-)

Sascha-Peter replied:
Okay, though it's related to the four points you mention straight after.

LukeCarrier commented on _docs/README.md:
```diff
@@ -27,6 +27,8 @@ The technical specifications of the environment are as follows:
 * [PostgreSQL 9.5](http://www.postgresql.org/)
 * [Selenium](http://www.seleniumhq.org/) hub and nodes.
+* Single sign on with [SimpleSAMLphp](https://simplesamlphp.org/)
+* Saml data store with [Redis](https://redis.io/)
```
Can we make this more generic? "Caching with Redis" or similar? Would also be good to include the major version number.

Sascha-Peter replied:
Sounds better indeed. Thanks, added it in.

LukeCarrier commented on _docs/SUMMARY.md:
```diff
@@ -11,6 +11,8 @@
         * [Default release](roles/app/default-release.md)
         * [Error pages](roles/app/error-pages.md)
         * [GoCD agent](roles/app/gocd-agent.md)
+        * [SAML](roles/app/saml.md)
```
The `app` documentation will need the `app-moodle` role splitting, too.

Sascha-Peter replied:
Split out as we discussed. 
Wasn't sure if the debug or error pages role should be moved to moodle only though :thinking:

LukeCarrier commented on _docs/roles/redis.md:
```diff
@@ -0,0 +1,28 @@
+
+## Redis Setup
+
+The supporting [redis](https://redis.io) setup for this SSO solution consists of a master and two slave instances.
```
Again, Redis won't be specific to SimpleSAMLphp -- describe the role in isolation here.

Sascha-Peter replied:
Changed.

LukeCarrier commented on _docs/roles/redis.md:
```diff
@@ -0,0 +1,28 @@
+- Place the `redis.conf` in `/etc/redis/`
+- Reload the service
+
+If the role is executed on a slave instance, it will configure the master-slave replication.
```
How? This is not how Redis Sentinel works, and has not been implemented in the states as far as I can see.

Sascha-Peter replied:
I removed the sentinel part for now and documented how it is right now. 
I will adjust once the sentinel work is done.

LukeCarrier replied:
:+1:

LukeCarrier commented on _docker/salt/grain/ci:
```diff
@@ -1,6 +1,6 @@
 environment: integration
 roles:
-  - app
+  - app-moodle
```
[Looks like we still need to do some work to get the Docker container running.](https://hub.docker.com/r/lukecarrier/moodle-ubiquitous/builds/bztwcc8igjeuqtgvjabhavi/) I know you've not had to interact with this and am happy to remedy it if necessary? It'll definitely be pillar configuration, as we already have the same states functioning under Vagrant.

Sascha-Peter replied:
That would be appreciated. Thanks.

LukeCarrier commented on _docs/roles/README.md:
```diff
@@ -38,34 +40,42 @@ Application servers provide the frontend of the environment, serving the site th
 | Dependencies | Dependants | Conflicts |
 | --- | --- | --- |
-| `app` | None | None |
+| `app-base` | None | None |
```
This section needs a little rework -- `app-moodle-debug` to configure the Behat stuff, `app-debug` to configure more generic PHP/xdebug stuff.

Sascha-Peter replied:
Adjusted the `app-debug` entry and expanded with `app-moodle-debug`

LukeCarrier commented on _docs/roles/README.md:
```diff
@@ -38,34 +40,42 @@ Application servers provide the frontend of the environment, serving the site th
 | Dependencies | Dependants | Conflicts |
 | --- | --- | --- |
-| `app` | None | `app-gocd-agent` |
+| `app-base` | None | `app-gocd-agent` |
```
Don't forget `app-moodle-default-release`! :smile:

LukeCarrier commented on _docs/roles/app-saml/saml.md:
```diff
@@ -0,0 +1,57 @@
+# SSO
```
Nitpicking, but could this title match the one on the roles page?

Sascha-Peter replied:
Done, put in line with the others.

LukeCarrier commented on _docs/roles/app-saml/saml.md:
```diff
@@ -0,0 +1,57 @@
+- `metadata/saml20-sp-remote.php`
+- `enable` files to enable the exampleauth and redis modules
+  - `modules/exampleauth/enable`
+  - `modules/redis/enable`
```
Please use asterisks for bullet points with 4 space tabs for indentation. [EditorConfig](http://editorconfig.org/) should help normalise this if you install the plugin for your editor.

Sascha-Peter replied:
Done and done on the rest of the file.

LukeCarrier commented on _docs/roles/app-saml/saml.md:
```diff
@@ -0,0 +1,57 @@
+- Login
+- FirstName
+- LastName
+- Email
```
Since these are claim names, could we use code tags (single backtick, e.g. ``* `Login` ``)?

Sascha-Peter replied:
Done.

LukeCarrier commented on _docs/roles/redis.md:
```diff
@@ -0,0 +1,19 @@
+
+## Redis Setup
+
+The `redis` role will set up [Redis](https://redis.io) with a master and two slave servers.
```
The number of servers is down to the user (they have to add the desired number of servers with the `roles` grain then configure them using the pillar. Could we rephrase this to maybe:

> The `redis` role will set up [Redis](https://redis.io), optionally in a master and slave configuration.

Sascha-Peter replied:
Rephrased.

LukeCarrier commented on _docs/roles/app-saml/saml.md:
```diff
@@ -0,0 +1,57 @@
+- `config/authsources.php`
+- identity provider ssl `.crt` and `.pem` under `cert/`
+- `metadata/saml20-sp-remote.php`
+- `modules/exampleauth/enalbe`
```
Asterisks for bullets please :wink:

Sascha-Peter replied:
See above / done.

LukeCarrier commented on _docs/SUMMARY.md:
```diff
@@ -11,6 +11,8 @@
         * [Error pages](roles/app/error-pages.md)
         * [GoCD agent](roles/app/gocd-agent.md)
+        * [SAML](roles/app-saml/saml.md)
+    * [Redis](roles/redis.md)
```
Don't forget to add a link to the Moodle role page here too.

Sascha-Peter replied:
Updated.

LukeCarrier commented on _docs/roles/redis.md:
```diff
@@ -0,0 +1,19 @@
+- Configure an iptables rule for inbound traffic
+- Set `vm.overcommit_memory` to `1`
+- Place the `redis.conf` in `/etc/redis/`
+- Reload the service
```
Asterisks please! (sorry)

Sascha-Peter replied:
Done.

LukeCarrier commented on _vagrant/salt/pillar/platforms-saml-identity-provider.sls:
```diff
@@ -0,0 +1,147 @@
+      test_user_login_02: testavd02
+      test_user_firstname_02: Avado2
+      test_user_lastname_02: Avocado2
+      test_user_email_02: sascha.peter@fusetech.io
```
We'll need to make this more configurable -- having a fixed number of users isn't ideal, be we can also better expose SimpleSAMLphp's authentication sources here (you can create multiple instances, and then use those names elsewhere -- this is how you configure sources for multiple identity providers).

Since we're directly exposing the PHP configuration files elsewhere, could we amend this so that instead of the `test_user` keys you can directly enter the contents of the `authsources.php` file, maybe in a key named `auth_sources`?

Sascha-Peter replied:
I've restructured the pillar and made the code iterate over the values. Will adjust the vagrant config for that as well shortly.

Sascha-Peter replied:
Adjusted the vagrant file as well to make it flexible.

LukeCarrier commented on app-base/init.sls:
```diff
@@ -386,47 +346,28 @@ app.{{ domain }}.{{ instance }}.php-fpm:
+
+app.nginx.restart:
+  cmd.run:
+    - name: systemctl restart nginx
```
Is there a reason we don't attempt a `reload` here? I originally added in the restarts as a way of recovering from nginx not starting because e.g. it had no virtualhosts -- in all cases but changing bindings (e.g. when first enabling SSL and binding to port 443) a reload is sufficient to change the configuration.

Sascha-Peter replied:
None other than that this is what we had in the original role with a reload and a restart thereafter. Happy to remove what was on `master`.

LukeCarrier commented on app-moodle-debug/init.sls:
```diff
@@ -0,0 +1,46 @@
+#
+
+include:
+- app-debug
```
`app-moodle` should be included above `app-debug` here as a bunch of the requisites for these states won't be ready yet without it.

Sascha-Peter replied:
Done.

LukeCarrier commented on app-moodle-debug/init.sls:
```diff
@@ -0,0 +1,46 @@
+    - mode: 0644
+{% if pillar['systemd']['apply'] %}
+    - onchanges_in:
+      - app-debug.nginx.reload
```
This requisite will need duplicating into this state to ensure that the restart takes place reliably (`app-moodle-debug.nginx.reload`)

Sascha-Peter replied:
Done.

LukeCarrier commented on app-moodle-default-release/init.sls:
```diff
@@ -0,0 +1,21 @@
+#
+
+include:
+  - app-default-release
```
Missing the `app-moodle` include above `app-default-release` here.

Sascha-Peter replied:
Adjusted.

Sascha-Peter replied:
I wonder how the role worked before the separation as it was missing it :thinking:

LukeCarrier commented on app-moodle/init.sls:
```diff
@@ -0,0 +1,85 @@
+    - require:
+      - file: moodle.{{ domain }}.home
+
+moodle..{{ domain }}.localcache:
```
Duplicate `.` in the state name here

Sascha-Peter replied:
Done.

LukeCarrier commented on app-saml/config/authsources.php:
```diff
@@ -0,0 +1,396 @@
+<?php
```
I suggest removing this template and writing the file out directly from the pillar as per the comment there. Its content is largely default and the majority of the other files have been exposed this way, so it seems inconsistent to have such a different template for this file.

Sascha-Peter replied:
I might be looking in the wrong place but I can't seem to find the comments you're referencing here in the pillar repo. is it on the saml branch?

Sascha-Peter replied:
I've removed the unnecessary comments down below and also may re-jigg the two templates a bit to merge them into one, since the only difference (at the moment) is on the auth plugin. Will add in the configuration option for the plugins as well and ping you when I finalise that. Not sure if I can move this entirely into the pillars but will see what I can do.

LukeCarrier commented on app-saml/init.sls:
```diff
@@ -0,0 +1,209 @@
+#
```
We'll need to give some thought to where we place these configuration files -- we shouldn't be templating them directly into the SimpleSAMLphp configuration directories as they may not exist yet and we generally want to manage configuration via deployments in GoCD. We'll review this later today?

Sascha-Peter replied:
Changed to put all of the configuration related files into a `conf` folder during the state execution and added the relevant scripts to tie into gocd and move the configuration files into the right place :+1:

LukeCarrier commented on app-saml/saml/idp-authsources.php.jinja:
```diff
@@ -0,0 +1,53 @@
+<?php
```
As per the comment in the other authsources.php file, is there a way we can expose more of this configuration to the user and remove this template?

Sascha-Peter commented:
Updated the pull request with all changes requested / suggestions incorporated.

Additionally I have rewritten the salt states to iterate through pillar values rather than having the expected files hard coded in the states as per our offline discussion. 

Happy for you to review and I'll gladly adjust any changes coming from this :octocat:

LukeCarrier commented:
Thanks for all the effort on these @Sascha-Peter.

These changes have been merged to `master` following a rebase to group related work items and the following changes:

* The `app-base` role now only managed server-level configuration. Everything else has been moved into a `app_platform` macro which is rendered in the `app-moodle` and `app-saml` roles. This allows us to cut down on the number of service restarts by not duplicating them across states (`include` directives don't allow you to add inverse requisites as the included states execute before the one containing the `include` directive).
* Simplified the Redis configuration by boiling it down to type juggling, since the configuration schema tends to change a lot across versions and I wanted to cut down on the amount of comments.
* Added support for configuring Redis Sentinel via the pillar.
* Removed the Redis VMs from the Vagrant configuration.
* Fixed a number of issues related to cipher and NameID negotiation with remote IdPs.
* Reworked SSL to reduce duplication in the per-platform nginx configuration and further decouple management of certificates with Lets Encrypt from nginx configuration.
* Added an `ssl-local` state to handle dumping certificates obtained externally to the disk.
